docker/development/Dockerfile

```diff
@@ -7,27 +7,63 @@ RUN GOOS=js GOARCH=wasm go build -o pdfcomprezzor.wasm
 RUN cp "$(go env GOROOT)/misc/wasm/wasm_exec.js" .
 
 # Now build the actual mampf application
-FROM ruby:3.1.4
+# https://hub.docker.com/_/ruby/
+FROM ruby:3.1.4-bullseye
 ENV RAILS_ENV=production
+
 EXPOSE 3000
 EXPOSE 9394
 
-# https://github.com/nodesource/distributions#installation-instructions
-RUN curl -sL https://deb.nodesource.com/setup_14.x | bash - && \
-    curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - && echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list && \
-    curl -sS https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - && echo "deb http://apt.postgresql.org/pub/repos/apt/ `lsb_release -cs`-pgdg main" | tee  /etc/apt/sources.list.d/pgdg.list && \
-    apt-get update && \
-    apt-get install --no-install-recommends -y ffmpeg ghostscript imagemagick \
-            libarchive-tools nodejs pdftk postgresql-client-13 sqlite3 wget \
-            wait-for-it yarn shared-mime-info && \
-    rm -rf /var/lib/apt/lists/* && apt-get clean
+# use "--login" option, so that .bashrc is sourced in new shells
+SHELL ["/bin/bash", "--login", "-c"] 
+
+# Install Node.js
+# https://github.com/nodesource/distributions/issues/1583#issuecomment-1597489401
+# https://stackoverflow.com/a/57546198/
+# Unfortunately, we have to explicitly specify the node version here
+# and cannot use 16.x as we need to put the node binary into the PATH
+# and therefore require the exact version to find the folder
+ENV NODE_VERSION=16.20.1
+RUN curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.0/install.sh | bash
+ENV NVM_DIR=/root/.nvm
+RUN source "${NVM_DIR}/nvm.sh" && nvm install "${NODE_VERSION}" && \
+    nvm use "${NODE_VERSION}" && nvm alias default "${NODE_VERSION}"
+
+ENV NODE_PATH="${NVM_DIR}/versions/node/v${NODE_VERSION}/bin/"
+ENV PATH="${NODE_PATH}:${PATH}"
+
+RUN nvm current
+RUN node --version
+RUN npm --version
 
+# Install Yarn (see https://yarnpkg.com/getting-started/install)
+# https://github.com/nodejs/corepack#corepack-prepare--nameversion
+ENV YARN_VERSION=1.22.19
+RUN corepack enable
+RUN corepack prepare "yarn@${YARN_VERSION}" --activate
+RUN which yarn; yarn --version
+# even though this is not specified in the corepack documentation, we need to
+# run "set version", otherwise the app user will have an old version of
+# yarn avaialble instead of our specified version
+RUN yarn set version "${YARN_VERSION}"
```

Comment:
Group these `RUN` statements?

Reply:
Not having these `RUN` statements grouped will lead to a better experience when reading the logs in my opinion. Since when you group them, you get one statement like `RUN which yarn; yarn --version`, but then two lines of output and you have to match which line of the output belongs to which statement. While this is certainly fine for this simple RUN statement, it might get a bit messy when grouping all of these statements together.

And one more benefit: if you were to run `corepack enable` and `corepack prepare ...` in one layer/shell and the `corepack enable` statement fails, maybe it's not sure which of these commands really failed (although I hope that `corepack` and other binaries will print the exact command that failed but I wouldn't rely on that).

Additionally, I don't think (though not tested yet) that spinning up a new shell for multiple `RUN` statements is that much of a performance impact. For `production` it does not matter anyways, but I see that it does indeed matter for `development` as one wants to get the setup up and running quickly.